File: backend/services/stigmer-server/pkg/domain/agentexecution/temporal/dispatch_test.go
package temporal

import (
	"context"
	"testing"

	sessionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/session/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/backend/libs/go/store/sqlite"
)

func setupTestStore(t *testing.T) *sqlite.Store {
	t.Helper()
	s, err := sqlite.NewStore(t.TempDir() + "/dispatch_test.sqlite")
	if err != nil {
		t.Fatalf("failed to create store: %v", err)
	}
	return s
}

func saveSession(t *testing.T, s *sqlite.Store, id string, harness sessionv1.Harness) {
	t.Helper()
	session := &sessionv1.Session{
		ApiVersion: "agentic.stigmer.ai/v1",
		Kind:       "Session",
		Metadata:   &apiresource.ApiResourceMetadata{Id: id, Name: "test-session", Org: "test-org"},
		Spec:       &sessionv1.SessionSpec{AgentInstanceId: "test-instance", Harness: harness},
	}
	if err := s.SaveResource(context.Background(), apiresourcekind.ApiResourceKind_session, id, session); err != nil {
		t.Fatalf("failed to save session: %v", err)
	}
}

func globalConfig() *Config {
	return &Config{
		StigmerQueue:           "agent_execution_stigmer",
		RunnerQueue:            DefaultActivityTaskQueue,
		ActivityRouting:        RoutingGlobal,
		DefaultExecutionTarget: DefaultExecutionTargetLocal,
	}
}

func sessionConfig() *Config {
	return &Config{
		StigmerQueue:           "agent_execution_stigmer",
		RunnerQueue:            DefaultActivityTaskQueue,
		ActivityRouting:        RoutingSession,
		DefaultExecutionTarget: DefaultExecutionTargetLocal,
	}
}

func cloudConfig() *Config {
	return &Config{
		StigmerQueue:           "agent_execution_stigmer",
		RunnerQueue:            DefaultActivityTaskQueue,
		ActivityRouting:        RoutingSession,
		DefaultExecutionTarget: DefaultExecutionTargetCloud,
	}
}

func TestFormatSessionTaskQueue(t *testing.T) {
	tests := []struct {
		sessionID string
		want      string
	}{
		{"ses_01arz3ndektsv4rrffq69g5fav", "session:ses_01arz3ndektsv4rrffq69g5fav"},
		{"ses_abc123", "session:ses_abc123"},
		{"any-arbitrary-id", "session:any-arbitrary-id"},
	}

	for _, tt := range tests {
		t.Run(tt.sessionID, func(t *testing.T) {
			got := FormatSessionTaskQueue(tt.sessionID)
			if got != tt.want {
				t.Errorf("FormatSessionTaskQueue(%q) = %q, want %q", tt.sessionID, got, tt.want)
			}
		})
	}
}

func TestResolveActivityTaskQueue_GlobalRouting(t *testing.T) {
	cfg := globalConfig()

	t.Run("no session ID — returns default queue with NATIVE harness", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		result, err := ResolveActivityTaskQueue(context.Background(), s, "", cfg, "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.TaskQueue != DefaultActivityTaskQueue {
			t.Errorf("expected task queue %q, got %q", DefaultActivityTaskQueue, result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_NATIVE {
			t.Errorf("expected HARNESS_NATIVE, got %v", result.Harness)
		}
	})

	t.Run("session not found — returns default queue without error", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		result, err := ResolveActivityTaskQueue(context.Background(), s, "nonexistent-session", cfg, "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.TaskQueue != DefaultActivityTaskQueue {
			t.Errorf("expected task queue %q, got %q", DefaultActivityTaskQueue, result.TaskQueue)
		}
	})

	t.Run("valid session — returns default queue even with session present", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		saveSession(t, s, "ses_global_test", sessionv1.Harness_HARNESS_CURSOR)

		result, err := ResolveActivityTaskQueue(context.Background(), s, "ses_global_test", cfg, "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.TaskQueue != DefaultActivityTaskQueue {
			t.Errorf("global routing must always return default queue, got %q", result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_CURSOR {
			t.Errorf("expected HARNESS_CURSOR, got %v", result.Harness)
		}
	})
}

func TestResolveActivityTaskQueue_SessionRouting(t *testing.T) {
	cfg := sessionConfig()

	t.Run("no session ID — falls back to default queue", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		result, err := ResolveActivityTaskQueue(context.Background(), s, "", cfg, "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.TaskQueue != DefaultActivityTaskQueue {
			t.Errorf("empty session ID must fall back to default queue, got %q", result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_NATIVE {
			t.Errorf("expected HARNESS_NATIVE, got %v", result.Harness)
		}
	})

	t.Run("valid session ID — returns per-session queue", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_01arz3ndektsv4rrffq69g5fav"
		saveSession(t, s, sessionID, sessionv1.Harness_HARNESS_NATIVE)

		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, cfg, "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		expectedQueue := "session:" + sessionID
		if result.TaskQueue != expectedQueue {
			t.Errorf("expected task queue %q, got %q", expectedQueue, result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_NATIVE {
			t.Errorf("expected HARNESS_NATIVE, got %v", result.Harness)
		}
	})

	t.Run("session not found — still returns per-session queue", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_nonexistent"
		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, cfg, "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		expectedQueue := "session:" + sessionID
		if result.TaskQueue != expectedQueue {
			t.Errorf("expected task queue %q, got %q", expectedQueue, result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_NATIVE {
			t.Errorf("expected HARNESS_NATIVE (default when session not found), got %v", result.Harness)
		}
	})

	t.Run("CURSOR harness — returns per-session queue with correct harness", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_cursor_session"
		saveSession(t, s, sessionID, sessionv1.Harness_HARNESS_CURSOR)

		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, cfg, "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		expectedQueue := "session:" + sessionID
		if result.TaskQueue != expectedQueue {
			t.Errorf("expected task queue %q, got %q", expectedQueue, result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_CURSOR {
			t.Errorf("expected HARNESS_CURSOR, got %v", result.Harness)
		}
	})
}

func TestResolveTaskQueue(t *testing.T) {
	tests := []struct {
		name      string
		sessionID string
		cfg       *Config
		want      string
	}{
		{
			name:      "global routing ignores session ID",
			sessionID: "ses_123",
			cfg:       globalConfig(),
			want:      DefaultActivityTaskQueue,
		},
		{
			name:      "global routing with empty session ID",
			sessionID: "",
			cfg:       globalConfig(),
			want:      DefaultActivityTaskQueue,
		},
		{
			name:      "session routing with valid ID",
			sessionID: "ses_abc",
			cfg:       sessionConfig(),
			want:      "session:ses_abc",
		},
		{
			name:      "session routing with empty ID falls back",
			sessionID: "",
			cfg:       sessionConfig(),
			want:      DefaultActivityTaskQueue,
		},
		{
			name:      "custom runner queue used as fallback",
			sessionID: "",
			cfg:       &Config{RunnerQueue: "custom_queue", ActivityRouting: RoutingSession},
			want:      "custom_queue",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := resolveTaskQueue(tt.sessionID, tt.cfg)
			if got != tt.want {
				t.Errorf("resolveTaskQueue(%q) = %q, want %q", tt.sessionID, got, tt.want)
			}
		})
	}
}

func TestResolveExecutionTarget(t *testing.T) {
	tests := []struct {
		name   string
		target sessionv1.ExecutionTarget
		cfg    *Config
		want   sessionv1.ExecutionTarget
	}{
		{
			name:   "LOCAL target passes through",
			target: sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL,
			cfg:    globalConfig(),
			want:   sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL,
		},
		{
			name:   "CLOUD target passes through",
			target: sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD,
			cfg:    globalConfig(),
			want:   sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD,
		},
		{
			name:   "UNSPECIFIED resolves to LOCAL when default is local",
			target: sessionv1.ExecutionTarget_EXECUTION_TARGET_UNSPECIFIED,
			cfg:    sessionConfig(),
			want:   sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL,
		},
		{
			name:   "UNSPECIFIED resolves to CLOUD when default is cloud",
			target: sessionv1.ExecutionTarget_EXECUTION_TARGET_UNSPECIFIED,
			cfg:    cloudConfig(),
			want:   sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.cfg.ResolveExecutionTarget(tt.target)
			if got != tt.want {
				t.Errorf("Config.ResolveExecutionTarget(%v) = %v, want %v", tt.target, got, tt.want)
			}
		})
	}
}

func TestResolveActivityTaskQueue_ExecutionTarget(t *testing.T) {
	t.Run("session with LOCAL target returns LOCAL in result", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_local"
		session := &sessionv1.Session{
			ApiVersion: "agentic.stigmer.ai/v1",
			Kind:       "Session",
			Metadata:   &apiresource.ApiResourceMetadata{Id: sessionID, Name: "test", Org: "test-org"},
			Spec: &sessionv1.SessionSpec{
				AgentInstanceId: "test-instance",
				Harness:         sessionv1.Harness_HARNESS_NATIVE,
				ExecutionTarget: sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL,
			},
		}
		if err := s.SaveResource(context.Background(), apiresourcekind.ApiResourceKind_session, sessionID, session); err != nil {
			t.Fatalf("failed to save session: %v", err)
		}

		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, sessionConfig(), "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL {
			t.Errorf("expected LOCAL, got %v", result.ExecutionTarget)
		}
	})

	t.Run("session with CLOUD target returns CLOUD in result", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_cloud"
		session := &sessionv1.Session{
			ApiVersion: "agentic.stigmer.ai/v1",
			Kind:       "Session",
			Metadata:   &apiresource.ApiResourceMetadata{Id: sessionID, Name: "test", Org: "test-org"},
			Spec: &sessionv1.SessionSpec{
				AgentInstanceId: "test-instance",
				Harness:         sessionv1.Harness_HARNESS_NATIVE,
				ExecutionTarget: sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD,
			},
		}
		if err := s.SaveResource(context.Background(), apiresourcekind.ApiResourceKind_session, sessionID, session); err != nil {
			t.Fatalf("failed to save session: %v", err)
		}

		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, sessionConfig(), "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD {
			t.Errorf("expected CLOUD, got %v", result.ExecutionTarget)
		}
	})

	t.Run("session with UNSPECIFIED target resolves based on config", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_unspecified"
		saveSession(t, s, sessionID, sessionv1.Harness_HARNESS_NATIVE)

		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, cloudConfig(), "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD {
			t.Errorf("expected CLOUD (from config default), got %v", result.ExecutionTarget)
		}
	})

	t.Run("activity_task_queue override routes to parent workflow sandbox", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_override_test"
		saveSession(t, s, sessionID, sessionv1.Harness_HARNESS_CURSOR)

		override := "wfexec:wfx_parent_abc123"
		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, sessionConfig(), override)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.TaskQueue != override {
			t.Errorf("expected override queue %q, got %q", override, result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_CURSOR {
			t.Errorf("expected session's harness CURSOR, got %v", result.Harness)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL {
			t.Errorf("expected LOCAL (no provisioning needed), got %v", result.ExecutionTarget)
		}
	})

	t.Run("activity_task_queue override with no session still works", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		override := "wfexec:wfx_no_session"
		result, err := ResolveActivityTaskQueue(context.Background(), s, "", cloudConfig(), override)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.TaskQueue != override {
			t.Errorf("expected override queue %q, got %q", override, result.TaskQueue)
		}
		if result.Harness != sessionv1.Harness_HARNESS_NATIVE {
			t.Errorf("expected default NATIVE harness, got %v", result.Harness)
		}
		if result.ExecutionTarget != sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL {
			t.Errorf("expected LOCAL (sandbox already exists), got %v", result.ExecutionTarget)
		}
	})

	t.Run("empty override falls through to normal routing", func(t *testing.T) {
		s := setupTestStore(t)
		defer s.Close()

		sessionID := "ses_no_override"
		saveSession(t, s, sessionID, sessionv1.Harness_HARNESS_NATIVE)

		result, err := ResolveActivityTaskQueue(context.Background(), s, sessionID, sessionConfig(), "")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result.TaskQueue != FormatSessionTaskQueue(sessionID) {
			t.Errorf("expected session queue %q, got %q", FormatSessionTaskQueue(sessionID), result.TaskQueue)
		}
	})
}

File: backend/services/stigmer-server/src/temporal/agentexecution/__tests__/dispatch.test.ts
/**
 * Dispatch-resolution tests — ports
 * pkg/domain/agentexecution/temporal/dispatch_test.go case-for-case: the
 * queue-routing modes (global vs session), the harness/execution-target
 * extraction with the session-not-found degrade, the UNSPECIFIED-target
 * resolution against the config default, and the sandbox-affinity
 * override lane (opaque queue pass-through with forced LOCAL target so
 * cloud sandbox provisioning never triggers, and NATIVE-harness default
 * when the session is absent). Go's private resolveTaskQueue cases are
 * asserted through resolveActivityTaskQueue — the public boundary is the
 * contract.
 */
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { create } from "@bufbuild/protobuf";
import { afterEach, describe, expect, it } from "vitest";

import { SessionSchema } from "@stigmer/protos/ai/stigmer/agentic/session/v1/api_pb";
import {
  ExecutionTarget,
  Harness,
} from "@stigmer/protos/ai/stigmer/agentic/session/v1/enum_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import { createLogger } from "../../../boot/logger.js";
import {
  AgentExecutionTemporalConfig,
  DEFAULT_EXECUTION_TARGET_CLOUD,
  DEFAULT_EXECUTION_TARGET_LOCAL,
  ROUTING_GLOBAL,
  ROUTING_SESSION,
} from "../../../domain/agentexecution/temporal/config.js";
import { SqliteStore } from "../../../store/sqlite/store.js";
import type { Store } from "../../../store/interface.js";
import { formatSessionTaskQueue, resolveActivityTaskQueue } from "../dispatch.js";
import { DEFAULT_ACTIVITY_TASK_QUEUE } from "../names.js";

const silentLogger = createLogger({
  level: "error",
  pretty: false,
  write: () => {},
});

const cleanups: Array<() => Promise<void> | void> = [];
afterEach(async () => {
  for (const cleanup of cleanups.splice(0)) {
    await cleanup();
  }
});

function newStore(): Store {
  const dir = mkdtempSync(path.join(tmpdir(), "dispatch-test-"));
  const store = SqliteStore.open(path.join(dir, "dispatch_test.sqlite"), silentLogger);
  cleanups.push(async () => {
    await store.close();
    rmSync(dir, { recursive: true, force: true });
  });
  return store;
}

async function saveSession(
  store: Store,
  id: string,
  harness: Harness,
  executionTarget = ExecutionTarget.UNSPECIFIED,
): Promise<void> {
  await store.saveResource(
    ApiResourceKind.session,
    id,
    SessionSchema,
    create(SessionSchema, {
      apiVersion: "agentic.stigmer.ai/v1",
      kind: "Session",
      metadata: { id, name: "test-session", org: "test-org" },
      spec: { agentInstanceId: "test-instance", harness, executionTarget },
    }),
  );
}

function globalConfig(): AgentExecutionTemporalConfig {
  return new AgentExecutionTemporalConfig(
    "agent_execution_stigmer",
    DEFAULT_ACTIVITY_TASK_QUEUE,
    ROUTING_GLOBAL,
    DEFAULT_EXECUTION_TARGET_LOCAL,
  );
}

function sessionConfig(): AgentExecutionTemporalConfig {
  return new AgentExecutionTemporalConfig(
    "agent_execution_stigmer",
    DEFAULT_ACTIVITY_TASK_QUEUE,
    ROUTING_SESSION,
    DEFAULT_EXECUTION_TARGET_LOCAL,
  );
}

function cloudConfig(): AgentExecutionTemporalConfig {
  return new AgentExecutionTemporalConfig(
    "agent_execution_stigmer",
    DEFAULT_ACTIVITY_TASK_QUEUE,
    ROUTING_SESSION,
    DEFAULT_EXECUTION_TARGET_CLOUD,
  );
}

describe("formatSessionTaskQueue", () => {
  it.each([
    ["ses_01arz3ndektsv4rrffq69g5fav", "session:ses_01arz3ndektsv4rrffq69g5fav"],
    ["ses_abc123", "session:ses_abc123"],
    ["any-arbitrary-id", "session:any-arbitrary-id"],
  ])("formats %s", (sessionId, want) => {
    expect(formatSessionTaskQueue(sessionId)).toBe(want);
  });
});

describe("resolveActivityTaskQueue — global routing", () => {
  it("no session ID — returns default queue with NATIVE harness", async () => {
    const store = newStore();
    const result = await resolveActivityTaskQueue(store, "", globalConfig(), "", silentLogger);
    expect(result.taskQueue).toBe(DEFAULT_ACTIVITY_TASK_QUEUE);
    expect(result.harness).toBe(Harness.NATIVE);
  });

  it("session not found — returns default queue without error", async () => {
    const store = newStore();
    const result = await resolveActivityTaskQueue(
      store,
      "nonexistent-session",
      globalConfig(),
      "",
      silentLogger,
    );
    expect(result.taskQueue).toBe(DEFAULT_ACTIVITY_TASK_QUEUE);
  });

  it("valid session — returns default queue even with session present", async () => {
    const store = newStore();
    await saveSession(store, "ses_global_test", Harness.CURSOR);
    const result = await resolveActivityTaskQueue(
      store,
      "ses_global_test",
      globalConfig(),
      "",
      silentLogger,
    );
    expect(result.taskQueue, "global routing must always return default queue").toBe(
      DEFAULT_ACTIVITY_TASK_QUEUE,
    );
    expect(result.harness).toBe(Harness.CURSOR);
  });
});

describe("resolveActivityTaskQueue — session routing", () => {
  it("no session ID — falls back to default queue", async () => {
    const store = newStore();
    const result = await resolveActivityTaskQueue(store, "", sessionConfig(), "", silentLogger);
    expect(result.taskQueue).toBe(DEFAULT_ACTIVITY_TASK_QUEUE);
    expect(result.harness).toBe(Harness.NATIVE);
  });

  it("valid session ID — returns per-session queue", async () => {
    const store = newStore();
    const sessionId = "ses_01arz3ndektsv4rrffq69g5fav";
    await saveSession(store, sessionId, Harness.NATIVE);
    const result = await resolveActivityTaskQueue(
      store,
      sessionId,
      sessionConfig(),
      "",
      silentLogger,
    );
    expect(result.taskQueue).toBe(`session:${sessionId}`);
    expect(result.harness).toBe(Harness.NATIVE);
  });

  it("session not found — still returns per-session queue with NATIVE default", async () => {
    const store = newStore();
    const sessionId = "ses_nonexistent";
    const result = await resolveActivityTaskQueue(
      store,
      sessionId,
      sessionConfig(),
      "",
      silentLogger,
    );
    expect(result.taskQueue).toBe(`session:${sessionId}`);
    expect(result.harness, "NATIVE default when session not found").toBe(Harness.NATIVE);
  });

  it("CURSOR harness — returns per-session queue with correct harness", async () => {
    const store = newStore();
    const sessionId = "ses_cursor_session";
    await saveSession(store, sessionId, Harness.CURSOR);
    const result = await resolveActivityTaskQueue(
      store,
      sessionId,
      sessionConfig(),
      "",
      silentLogger,
    );
    expect(result.taskQueue).toBe(`session:${sessionId}`);
    expect(result.harness).toBe(Harness.CURSOR);
  });

  it("custom runner queue used as the empty-session fallback", async () => {
    const store = newStore();
    const config = new AgentExecutionTemporalConfig(
      "agent_execution_stigmer",
      "custom_queue",
      ROUTING_SESSION,
      DEFAULT_EXECUTION_TARGET_LOCAL,
    );
    const result = await resolveActivityTaskQueue(store, "", config, "", silentLogger);
    expect(result.taskQueue).toBe("custom_queue");
  });
});

describe("resolveActivityTaskQueue — execution target", () => {
  it("session with LOCAL target returns LOCAL in result", async () => {
    const store = newStore();
    await saveSession(store, "ses_local", Harness.NATIVE, ExecutionTarget.LOCAL);
    const result = await resolveActivityTaskQueue(
      store,
      "ses_local",
      sessionConfig(),
      "",
      silentLogger,
    );
    expect(result.executionTarget).toBe(ExecutionTarget.LOCAL);
  });

  it("session with CLOUD target returns CLOUD in result", async () => {
    const store = newStore();
    await saveSession(store, "ses_cloud", Harness.NATIVE, ExecutionTarget.CLOUD);
    const result = await resolveActivityTaskQueue(
      store,
      "ses_cloud",
      sessionConfig(),
      "",
      silentLogger,
    );
    expect(result.executionTarget).toBe(ExecutionTarget.CLOUD);
  });

  it("session with UNSPECIFIED target resolves based on config", async () => {
    const store = newStore();
    await saveSession(store, "ses_unspecified", Harness.NATIVE);
    const result = await resolveActivityTaskQueue(
      store,
      "ses_unspecified",
      cloudConfig(),
      "",
      silentLogger,
    );
    expect(result.executionTarget, "CLOUD from config default").toBe(ExecutionTarget.CLOUD);
  });

  it("activity_task_queue override routes to the parent workflow sandbox", async () => {
    const store = newStore();
    const sessionId = "ses_override_test";
    await saveSession(store, sessionId, Harness.CURSOR);
    const override = "wfexec:wfx_parent_abc123";
    const result = await resolveActivityTaskQueue(
      store,
      sessionId,
      sessionConfig(),
      override,
      silentLogger,
    );
    expect(result.taskQueue).toBe(override);
    expect(result.harness, "the session's harness still resolves").toBe(Harness.CURSOR);
    expect(result.executionTarget, "LOCAL — no provisioning needed").toBe(ExecutionTarget.LOCAL);
  });

  it("activity_task_queue override with no session still works", async () => {
    const store = newStore();
    const override = "wfexec:wfx_no_session";
    const result = await resolveActivityTaskQueue(
      store,
      "",
      cloudConfig(),
      override,
      silentLogger,
    );
    expect(result.taskQueue).toBe(override);
    expect(result.harness, "default NATIVE harness").toBe(Harness.NATIVE);
    expect(result.executionTarget, "LOCAL — sandbox already exists").toBe(ExecutionTarget.LOCAL);
  });

  it("surfaces non-NotFound store failures with the pinned dispatch message", async () => {
    const store = newStore();
    await store.close();
    // The message text is contract: the create step maps it VERBATIM to
    // FailedPrecondition (Go's ResolveActivityTaskQueue boundary).
    await expect(
      resolveActivityTaskQueue(store, "ses_any", sessionConfig(), "", silentLogger),
    ).rejects.toThrow(/^failed to load session for dispatch: /);
  });

  it("empty override falls through to normal routing", async () => {
    const store = newStore();
    const sessionId = "ses_no_override";
    await saveSession(store, sessionId, Harness.NATIVE);
    const result = await resolveActivityTaskQueue(
      store,
      sessionId,
      sessionConfig(),
      "",
      silentLogger,
    );
    expect(result.taskQueue).toBe(formatSessionTaskQueue(sessionId));
  });
});

describe("AgentExecutionTemporalConfig.resolveExecutionTarget", () => {
  it.each([
    ["LOCAL passes through", ExecutionTarget.LOCAL, globalConfig(), ExecutionTarget.LOCAL],
    ["CLOUD passes through", ExecutionTarget.CLOUD, globalConfig(), ExecutionTarget.CLOUD],
    [
      "UNSPECIFIED resolves to LOCAL when default is local",
      ExecutionTarget.UNSPECIFIED,
      sessionConfig(),
      ExecutionTarget.LOCAL,
    ],
    [
      "UNSPECIFIED resolves to CLOUD when default is cloud",
      ExecutionTarget.UNSPECIFIED,
      cloudConfig(),
      ExecutionTarget.CLOUD,
    ],
  ] as const)("%s", (_name, target, config, want) => {
    expect(config.resolveExecutionTarget(target)).toBe(want);
  });
});
